File: compiler/test/tests/cobalt/Parser/AssignParserTest.hs
module Parser.AssignParserTest where

import Test.HUnit

import TestUtil.ParserTestUtil
import AST.AST
import Parser.Parser

testAssignParser :: Test
testAssignParser = do
    let codeInlineNoType = "let x = y"
    let testInlineNoType = testParseSuccess codeInlineNoType (Assign (Name "x") Nothing True (Inline $ (Identifier (Name "y")))) statementParser

    let codeInlineNoTypeMutable = "let mutable x = y"
    let testInlineNoTypeMutable = testParseSuccess codeInlineNoTypeMutable (Assign (Name "x") Nothing False (Inline $ (Identifier (Name "y")))) statementParser

    let codeInline = "let x: Int = y"
    let testInline = testParseSuccess codeInline (Assign (Name "x") (Just (TypeRef (RefLocal (Name "Int")))) True (Inline $ (Identifier (Name "y")))) statementParser

    let codeInlineValue = "let x = 10"
    let testInlineValue = testParseSuccess codeInlineValue (Assign (Name "x") Nothing True (Inline (IntConst 10))) statementParser

    let codeDoBlock = unlines [ "let x: Int = do"
                              , "    x"
                              , "    y"
                              ]
    let testDoBlock = testParseSuccess codeDoBlock (Assign (Name "x") (Just (TypeRef (RefLocal (Name "Int")))) True (DoBlock $ BlockStmt [ExprAsStmt (Identifier (Name "x")),ExprAsStmt (Identifier (Name "y"))])) statementParser

    TestList [ testInlineNoType
             , testInlineNoTypeMutable
             , testInline
             , testInlineValue
             , testDoBlock
             ]

testAssignParserMultiple :: Test
testAssignParserMultiple = do
    let codeInlineNoType = "let x,y = z"
    let testInlineNoType = testParseSuccess codeInlineNoType (AssignMultiple [Name "x",Name "y"] Nothing True (Inline (Identifier (Name "z")))) statementParser

    let codeInlineNoTypeMutable = "let mutable x,y = z"
    let testInlineNoTypeMutable = testParseSuccess codeInlineNoTypeMutable (AssignMultiple [Name "x",Name "y"] Nothing False (Inline (Identifier (Name "z")))) statementParser

    let codeInline = "let x,y: Int = z"
    let testInline = testParseSuccess codeInline (AssignMultiple [Name "x",Name "y"] (Just (TypeRef (RefLocal (Name "Int")))) True (Inline (Identifier (Name "z")))) statementParser

    let codeDoBlockNoType = unlines [ "let x,y = do"
                              , "    i"
                              , "    j"
                              ]
    let testDoBlockNoType = testParseSuccess codeDoBlockNoType (AssignMultiple [Name "x",Name "y"] Nothing True (DoBlock $ BlockStmt [ExprAsStmt (Identifier (Name "i")),ExprAsStmt (Identifier (Name "j"))])) statementParser

    let codeDoBlock = unlines [ "let x,y: Int = do"
                              , "    i"
                              , "    j"
                              ]
    let testDoBlock = testParseSuccess codeDoBlock (AssignMultiple [Name "x",Name "y"] (Just (TypeRef (RefLocal (Name "Int")))) True (DoBlock $ BlockStmt [ExprAsStmt (Identifier (Name "i")),ExprAsStmt (Identifier (Name "j"))])) statementParser

    TestList [ testInlineNoType
             , testInlineNoTypeMutable
             , testInline
             , testDoBlockNoType
             , testDoBlock
             ]

File: src/test/scala/cobalt/parser/statement/AssignParserTest.scala
package cobalt.parser.statement

import cobalt.ast.AST._
import cobalt.parser.StatementParser
import cobalt.utils.TestUtil
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.{FunSpec, Matchers}

import scala.collection.mutable.ArrayBuffer

@RunWith(classOf[JUnitRunner])
class AssignParserTest extends FunSpec with Matchers
{
  describe("Assignment parser")
  {
    it("Should parse assignment")
    {
      TestUtil.parse("let x = 10", StatementParser.statement) shouldBe Assign(Name("x"),None,true,Inline(IntConst(10)))
    }

    it("Should parse mutable assignment")
    {
      TestUtil.parse("let mutable x = 10", StatementParser.statement) shouldBe Assign(Name("x"),None,false,Inline(IntConst(10)))
    }

    it("Should parse with type defined")
    {
      TestUtil.parse("let x: Int = 10", StatementParser.statement) shouldBe Assign(Name("x"),Some(TypeRef(RefLocal(Name("Int")))),true,Inline(IntConst(10)))
    }

    it("Should parse with a do block")
    {
      val code =
        """let x = do
          |  x
          |  y
        """.stripMargin.replace("\r", "")
      TestUtil.parse(code, StatementParser.statement) shouldBe Assign(Name("x"),None,true,DoBlock(BlockStmt(ArrayBuffer(ExprAsStmt(Identifier(Name("x"))), ExprAsStmt(Identifier(Name("y")))))))
    }
  }

  describe("Multiple assignment parser") {
    // TODO Assign parser multiple

    // TODO "let x,y = z"

    // TODO "let mutable x,y = z"

    // TODO "let x,y: Int = z"

    // TODO "let x,y = do"
    //      "    i"
    //      "    j"

    // TODO "let x,y: Int = do"
    //      "    i"
    //      "    j"
  }
}
